File: src/lib/spotify.js
let access_token = "";

export let currentSong = {};

async function getAccessToken() {
  const refresh_token = process.env.SPOTIFY_REFRESH_TOKEN;

  const response = await fetch("https://accounts.spotify.com/api/token", {
    method: "POST",
    headers: {
      Authorization: `Basic ${Buffer.from(
        `${process.env.SPOTIFY_CLIENT_ID}:${process.env.SPOTIFY_CLIENT_SECRET}`,
      ).toString("base64")}`,
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: new URLSearchParams({
      grant_type: "refresh_token",
      refresh_token,
    }),
  });
  return response.json();
}

let _spotifyCache = {
  timestamp: 0,
  text: null,
  status: null,
  headers: {},
};

export const currentlyPlayingSong = async () => {
  const now = Date.now();

  // Return cached response if it's younger than 10 seconds
  if (_spotifyCache.text !== null && now - _spotifyCache.timestamp < 10_000) {
    //console.log("Using cached Spotify response");
    return new Response(_spotifyCache.text, {
      status: _spotifyCache.status ?? 200,
      headers: _spotifyCache.headers,
    });
  }

  let response = await fetch(
    "https://api.spotify.com/v1/me/player/currently-playing",
    {
      headers: {
        Authorization: `Bearer ${access_token}`,
      },
      next: { revalidate: 10 },
    },
  );

  // If we didn't get a successful 200, try to refresh the access token and retry
  if (response.status !== 200) {
    const token = await getAccessToken();
    access_token = token.access_token;

    response = await fetch(
      "https://api.spotify.com/v1/me/player/currently-playing",
      {
        headers: {
          Authorization: `Bearer ${access_token}`,
        },
        next: { revalidate: 10 },
      },
    );
  }

  if (response.status === 204) {
    return { status: 204 };
  }

  // Clone and read response body to populate cache
  const cloned = response.clone();
  const text = await cloned.text();
  const headersObj = {};
  cloned.headers.forEach((value, key) => {
    headersObj[key] = value;
  });

  _spotifyCache = {
    timestamp: Date.now(),
    text,
    status: response.status,
    headers: headersObj,
  };

  return response;
};

File: src/spotify.rs
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};
use serde::{Serialize, Deserialize};
use reqwest::header::{AUTHORIZATION, CONTENT_TYPE};
use base64::{Engine as _, engine::general_purpose};

lazy_static::lazy_static! {
    static ref SPOTIFY_CACHE: Mutex<Option<CachedResponse>> = Mutex::new(None);
    static ref ACCESS_TOKEN: Mutex<String> = Mutex::new(String::new());
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct SpotifyResponse {
    pub artist: String,
    #[serde(rename = "isPlaying")]
    pub is_playing: bool,
    #[serde(rename = "songUrl")]
    pub song_url: String,
    pub title: String,
}

#[derive(Clone)]
struct CachedResponse {
    timestamp: u64,
    response: SpotifyResponse,
}

#[derive(Deserialize)]
struct TokenResponse {
    access_token: String,
}

#[derive(Deserialize)]
struct SpotifyPlayerItem {
    name: String,
    artists: Vec<SpotifyArtist>,
    external_urls: SpotifyUrls,
}

#[derive(Deserialize)]
struct SpotifyArtist {
    name: String,
}

#[derive(Deserialize)]
struct SpotifyUrls {
    spotify: String,
}

#[derive(Deserialize)]
struct SpotifyCurrentlyPlaying {
    is_playing: bool,
    item: Option<SpotifyPlayerItem>,
}

fn get_current_secs() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs()
}

async fn get_access_token() -> Result<String, Box<dyn std::error::Error + Send + Sync>> {
    let _ = dotenvy::dotenv();
    let client_id = std::env::var("SPOTIFY_CLIENT_ID")
        .map_err(|_| "SPOTIFY_CLIENT_ID environment variable not found")?;
    let client_secret = std::env::var("SPOTIFY_CLIENT_SECRET")
        .map_err(|_| "SPOTIFY_CLIENT_SECRET environment variable not found")?;
    let refresh_token = std::env::var("SPOTIFY_REFRESH_TOKEN")
        .map_err(|_| "SPOTIFY_REFRESH_TOKEN environment variable not found")?;

    let client = reqwest::Client::new();
    let auth_header = format!(
        "Basic {}",
        general_purpose::STANDARD.encode(format!("{}:{}", client_id, client_secret))
    );

    let res = client
        .post("https://accounts.spotify.com/api/token")
        .header(AUTHORIZATION, auth_header)
        .header(CONTENT_TYPE, "application/x-www-form-urlencoded")
        .form(&[
            ("grant_type", "refresh_token"),
            ("refresh_token", &refresh_token),
        ])
        .send()
        .await?
        .json::<TokenResponse>()
        .await?;

    Ok(res.access_token)
}

pub async fn get_current_song() -> SpotifyResponse {
    let now = get_current_secs();

    // 1. Check cache first
    {
        if let Ok(cache_guard) = SPOTIFY_CACHE.lock() {
            if let Some(cached) = &*cache_guard {
                if now - cached.timestamp < 10 {
                    return cached.response.clone();
                }
            }
        }
    }

    // 2. Fetch from Spotify API
    match fetch_currently_playing_retry().await {
        Ok(song) => {
            // Update cache
            if let Ok(mut cache_guard) = SPOTIFY_CACHE.lock() {
                *cache_guard = Some(CachedResponse {
                    timestamp: now,
                    response: song.clone(),
                });
            }
            song
        }
        Err(e) => {
            eprintln!("Error fetching Spotify status: {}", e);
            SpotifyResponse {
                artist: String::new(),
                is_playing: false,
                song_url: String::new(),
                title: String::new(),
            }
        }
    }
}

async fn fetch_currently_playing_retry() -> Result<SpotifyResponse, Box<dyn std::error::Error + Send + Sync>> {
    let mut token = {
        if let Ok(t_guard) = ACCESS_TOKEN.lock() {
            t_guard.clone()
        } else {
            String::new()
        }
    };

    if token.is_empty() {
        token = get_access_token().await?;
        if let Ok(mut t_guard) = ACCESS_TOKEN.lock() {
            *t_guard = token.clone();
        }
    }

    let client = reqwest::Client::new();
    let mut res = client
        .get("https://api.spotify.com/v1/me/player/currently-playing")
        .header(AUTHORIZATION, format!("Bearer {}", token))
        .send()
        .await?;

    // Retry once with refreshed token if auth fails (401 status)
    if res.status() == reqwest::StatusCode::UNAUTHORIZED {
        token = get_access_token().await?;
        if let Ok(mut t_guard) = ACCESS_TOKEN.lock() {
            *t_guard = token.clone();
        }
        res = client
            .get("https://api.spotify.com/v1/me/player/currently-playing")
            .header(AUTHORIZATION, format!("Bearer {}", token))
            .send()
            .await?;
    }

    if res.status() == reqwest::StatusCode::NO_CONTENT {
        return Ok(SpotifyResponse {
            artist: String::new(),
            is_playing: false,
            song_url: String::new(),
            title: String::new(),
        });
    }

    if !res.status().is_success() {
        return Err(format!("Spotify API returned status {}", res.status()).into());
    }

    let playing_data = res.json::<SpotifyCurrentlyPlaying>().await?;
    
    if let Some(item) = playing_data.item {
        let artists: Vec<String> = item.artists.into_iter().map(|a| a.name).collect();
        let artist_str = artists.join(", ");
        Ok(SpotifyResponse {
            artist: artist_str,
            is_playing: playing_data.is_playing,
            song_url: item.external_urls.spotify,
            title: item.name,
        })
    } else {
        Ok(SpotifyResponse {
            artist: String::new(),
            is_playing: false,
            song_url: String::new(),
            title: String::new(),
        })
    }
}
